tests/test_odp_segment_manager.py

```diff
@@ -0,0 +1,210 @@
+
+            odp_config = OdpConfig(self.api_key, self.api_host, [])
+            segments_cache = LRUCache(1000, 1000)
+            api = ZaiusGraphQLApiManager()
+
+            segment_manager = OdpSegmentManager(odp_config, segments_cache, api, mock_logger)
+            segments = segment_manager.fetch_qualified_segments(self.user_key, self.user_value, [])
+
+            self.assertEqual(segments, [])
+            mock_logger.debug.assert_called_once_with('No segments are used in the project. Returning empty list.')
+            mock_logger.error.assert_not_called()
+
+    def test_fetch_segments_success_cache_miss(self):
+        """
+        we are fetching user key/value 'fs_user_id'/'test-user-value'
+        which is different from what we have passed to cache (fs_user_id-$-123/['d'])
+        ---> hence we trigger a cache miss
+        """
+        with mock.patch('requests.post') as mock_request_post, \
+                mock.patch('optimizely.logger') as mock_logger:
+            mock_request_post.return_value = self.fake_server_response(status_code=200,
+                                                                       content=self.good_response_data)
+
+            odp_config = OdpConfig(self.api_key, self.api_host, ["a", "b", "c"])
+            segments_cache = LRUCache(1000, 1000)
+            api = ZaiusGraphQLApiManager()
+
+            segment_manager = OdpSegmentManager(odp_config, segments_cache, api, mock_logger)
+
+            cache_key = segment_manager.make_cache_key(self.user_key, '123')
+            segment_manager.segments_cache.save(cache_key, ["d"])
+
+            segments = segment_manager.fetch_qualified_segments(self.user_key, self.user_value, [])
+
+        self.assertEqual(segments, ["a", "b"])
+        actual_cache_key = segment_manager.make_cache_key(self.user_key, self.user_value)
+        self.assertEqual(segment_manager.segments_cache.lookup(actual_cache_key), ["a", "b"])
+        mock_logger.debug.assert_called_once_with('ODP cache miss. Making a call to ODP server.')
+        mock_logger.error.assert_not_called()
+
+    def test_fetch_segments_success_cache_hit(self):
+        with mock.patch('optimizely.logger') as mock_logger:
+            odp_config = OdpConfig()
+            odp_config.update(self.api_key, self.api_host, ['c'])
+            segments_cache = LRUCache(1000, 1000)
+            segment_manager = OdpSegmentManager(odp_config, segments_cache, None, mock_logger)
+
+            cache_key = segment_manager.make_cache_key(self.user_key, self.user_value)
+            segment_manager.segments_cache.save(cache_key, ['c'])
+
+            segments = segment_manager.fetch_qualified_segments(self.user_key, self.user_value, [])
+
+        self.assertEqual(segments, ['c'])
+        mock_logger.debug.assert_called_once_with('ODP cache hit. Returning segments from cache.')
+        mock_logger.error.assert_not_called()
+
```

Comment:
Can we also validate that graphQLApi.fetchSegment() is not called in this case?

Reply:
yes, check added, see line 98